File: sunrei-server/src/utils/polygon-utils.ts
export interface Point {
  latitude: number;
  longitude: number;
}

export function parseWKTPolygon(wkt: string): Point[] {
  const match = wkt.match(/POLYGON\s*\(\s*\(\s*([^)]+)\s*\)\s*\)/i);
  if (!match) {
    throw new Error('Invalid WKT polygon format');
  }
  
  const coordinatesStr = match[1];
  const points = coordinatesStr.split(',').map(coord => {
    const [lon, lat] = coord.trim().split(/\s+/).map(Number);
    return { latitude: lat, longitude: lon };
  });
  
  return points;
}

export function isPointInPolygon(point: Point, polygon: Point[]): boolean {
  let inside = false;
  
  for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    const xi = polygon[i].longitude;
    const yi = polygon[i].latitude;
    const xj = polygon[j].longitude;
    const yj = polygon[j].latitude;
    
    const intersect = ((yi > point.latitude) !== (yj > point.latitude))
        && (point.longitude < (xj - xi) * (point.latitude - yi) / (yj - yi) + xi);
    
    if (intersect) inside = !inside;
  }
  
  return inside;
}

File: sunrei-server/src/main/kotlin/com/sunrei/utils/PolygonUtils.kt
package com.sunrei.utils

data class Point(
    val latitude: Float,
    val longitude: Float
)

fun parseWKTPolygon(wkt: String): List<Point> {
    val regex = Regex("""POLYGON\s*\(\s*\(\s*([^)]+)\s*\)\s*\)""", RegexOption.IGNORE_CASE)
    val match = regex.find(wkt) ?: throw IllegalArgumentException("Invalid WKT polygon format")
    
    val coordinatesStr = match.groupValues[1]
    return coordinatesStr.split(',').map { coord ->
        val parts = coord.trim().split(Regex("\\s+"))
        val lon = parts[0].toFloat()
        val lat = parts[1].toFloat()
        Point(latitude = lat, longitude = lon)
    }
}

fun isPointInPolygon(point: Point, polygon: List<Point>): Boolean {
    var inside = false
    
    var j = polygon.size - 1
    for (i in polygon.indices) {
        val xi = polygon[i].longitude
        val yi = polygon[i].latitude
        val xj = polygon[j].longitude
        val yj = polygon[j].latitude
        
        val intersect = ((yi > point.latitude) != (yj > point.latitude)) &&
                (point.longitude < (xj - xi) * (point.latitude - yi) / (yj - yi) + xi)
        
        if (intersect) inside = !inside
        j = i
    }
    
    return inside
}